pkg/volume/persistent_claim/persistent_claim.go

```diff
@@ -0,0 +1,151 @@
+/*
+Copyright 2014 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package persistent_claim
+
+import (
+	"fmt"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/types"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume/gce_pd"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume/glusterfs"
+	"github.com/GoogleCloudPlatform/kubernetes/pkg/volume/host_path"
+	"github.com/golang/glog"
+)
+
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	return []volume.VolumePlugin{&persistentClaimPlugin{nil}}
+}
+
+type persistentClaimPlugin struct {
+	host volume.VolumeHost
+}
+
+var _ volume.VolumePlugin = &persistentClaimPlugin{}
+
+const (
+	persistentClaimPluginName = "kubernetes.io/persistent-claim"
+)
+
+func (plugin *persistentClaimPlugin) Init(host volume.VolumeHost) {
+	plugin.host = host
+}
+
+func (plugin *persistentClaimPlugin) Name() string {
+	return persistentClaimPluginName
+}
+
+func (plugin *persistentClaimPlugin) CanSupport(spec *api.Volume) bool {
+	return spec.PersistentVolumeClaimVolumeSource != nil
+}
+
+func (plugin *persistentClaimPlugin) NewBuilder(spec *api.Volume, podRef *api.ObjectReference) (volume.Builder, error) {
+	glog.V(5).Infof("Creating builder for %+v\n", spec)
+	claimName := spec.PersistentVolumeClaimVolumeSource.ClaimName
+	podNamespace := podRef.Namespace
+	claim, err := plugin.host.GetKubeClient().PersistentVolumeClaims(podNamespace).Get(claimName)
+
+	glog.V(5).Infof("Using claim: %+v\n", claim)
+
+	if err != nil {
+		glog.Errorf("Error finding claim in namespace  PersistentVolume by ClaimName: %+v\n", spec.PersistentVolumeClaimVolumeSource.ClaimName)
+		return nil, err
+	}
+
+	if claim != nil {
+		pv, err := plugin.host.GetKubeClient().PersistentVolumes().Get(claim.Status.VolumeRef.Name)
+		if err != nil {
+			glog.Errorf("Error finding bound PersistentVolume by ClaimName: %+v\n", spec.PersistentVolumeClaimVolumeSource.ClaimName)
+			return nil, err
+		}
+
+		glog.V(5).Infof("Using persistent volume: %+v\n", pv)
+		volPlugin := getVolumePlugin(pv.Spec)
```

Comment:
This is what VolumeHost.NewWrapperBuilder() is for.  See how Secret is implemented.  All you need to do is translate api.PersistentVolume to api.Volume (and frankly that could probably be reduced to PVSource and VSource in the Host APIs.